ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp

```diff
@@ -0,0 +1,405 @@
+
+        const std::string new_input_name = target_node.output(0) + ":" + edge.m_tensor_name;
+
+        if (is_graph_initializer(graph, edge.m_tensor_name))
+        {
+            replace_initializer_with_new_input(graph, edge, new_input_name);
+        }
+        else
+        {
+            auto& new_input = *(graph.add_input());
+            // copy the intermediate tensor properties to the newly created input
+            new_input.MergeFrom(find_tensor_descriptor(graph, edge.m_tensor_name));
+            *(new_input.mutable_name()) = new_input_name;
+        }
+
+        // attach the new graph input to the target node's input
+        *target_input = new_input_name;
+
+        return {true, InputEdge{edge.m_node_idx, new_input_name}};
+    }
+
+    void append_new_graph_output(ONNX_NAMESPACE::GraphProto& graph, const OutputEdge& edge)
+    {
+        if (already_exists(graph.output(), edge.m_tensor_name))
+        {
+            return;
+        }
+
+        auto& target_node = *(graph.mutable_node(edge.m_node_idx));
+        const auto& node_outputs = target_node.output();
+        const auto target_output =
+            std::find(node_outputs.begin(), node_outputs.end(), edge.m_tensor_name);
+
+        NGRAPH_CHECK(target_output != node_outputs.end(),
+                     "Output '",
+                     edge.m_tensor_name,
+                     "' not found in the outputs of node ",
+                     edge.m_node_idx,
+                     ". Cannot append a new graph output to this node.");
+
+        auto& new_output = *(graph.add_output());
+        // copy the intermediate tensor's properties to the newly created
+        new_output.MergeFrom(find_tensor_descriptor(graph, edge.m_tensor_name));
+        *(new_output.mutable_name()) = edge.m_tensor_name;
+    }
+
+    /// \brief Removes all items from a container except the ones whose names are in items_to_keep
+    ///        It's intended to work with ONNX graph inputs and initializers only.
+    template <typename Container>
+    void discard_by_name(Container& all_items, const std::set<std::string>& items_to_keep)
+    {
+        static_assert(
+            std::is_same<typename Container::value_type, ONNX_NAMESPACE::ValueInfoProto>::value ||
+            std::is_same<typename Container::value_type, ONNX_NAMESPACE::TensorProto>::value);
+
+        // The tested item can be discarded if its name is not found in the items_to_keep set
+        const auto can_be_discarded = [&items_to_keep](const typename Container::value_type& item) {
+            return items_to_keep.count(item.name()) == 0;
+        };
+
+        auto items_to_remove = all_items.size() - items_to_keep.size();
+
+        // move the elements-to-discard to the end of the container
+        auto new_end = all_items.end();
+        while (items_to_remove > 0)
+        {
+            new_end = std::remove_if(all_items.begin(), new_end, can_be_discarded);
```

Comment:
Why `all_items` have to be `remove_if` several times?

Reply:
This is a leftover from a previous approach, thanks for spotting it. I'll change it to a single call of remove_if.